ospackage/recipes/steps.go

```diff
@@ -84,39 +64,56 @@ func StepFileExec(step *osconfigpb.SoftwareRecipe_Step_FileExec, artifacts map[s
 		var ok bool
 		path, ok = artifacts[v.ArtifactId]
 		if !ok {
-			return nil, fmt.Errorf("%q not found in artifact map", v.ArtifactId)
+			return fmt.Errorf("%q not found in artifact map", v.ArtifactId)
 		}
 	default:
-		return nil, fmt.Errorf("can't determine location type")
+		return fmt.Errorf("can't determine location type")
 	}
 
-	res := []string{path}
-	res = append(res, step.FileExec.Args...)
-	return res, nil
+	return executeCommand(path, filepath.Join(runDir, "stepName"), runEnvs, step.FileExec.Args...)
 }
 
 // StepScriptRun builds the command for a ScriptRun step
-func StepScriptRun(step *osconfigpb.SoftwareRecipe_Step_ScriptRun, artifacts map[string]string) ([]string, error) {
+func StepScriptRun(step *osconfigpb.SoftwareRecipe_Step_ScriptRun, artifacts map[string]string, runEnvs []string, runDir string) error {
 	// TODO: should be putting this in stepN_type/ dir, but that needs me to know the dir way in advance..
 	// actually this is an artifact. we should have made them referenced as artifacts.. we'll have to repeat file creation logic here.
 	f, err := os.Create("/tmp/scriptrun")
 	if err != nil {
-		return nil, err
+		return err
 	}
 	defer f.Close()
 
 	f.WriteString(step.ScriptRun.Script)
 	f.Sync()
 	if err := os.Chmod("/tmp/scriptrun", 0755); err != nil {
-		return nil, err
+		return err
 	}
 
-	res := []string{"/bin/sh", "-c"}
-	//if step.ScriptRun.Interpreter == osconfigpb.SoftwareRecipe_Step_RunScript_POWERSHELL {
 	var qargs []string
+	//if step.ScriptRun.Interpreter == osconfigpb.SoftwareRecipe_Step_RunScript_POWERSHELL {
 	for _, arg := range step.ScriptRun.Args {
 		qargs = append(qargs, fmt.Sprintf("%q", arg))
 	}
-	res = append(res, "/tmp/scriptrun"+" "+strings.Join(qargs, " "))
-	return res, nil
+	command := "/tmp/scriptrun" + " " + strings.Join(qargs, " ")
+	args := []string{"-c", command}
+	return executeCommand("/bin/sh", filepath.Join(runDir, "stepName"), runEnvs, args...)
+}
+
+func executeCommand(cmd string, workDir string, envs []string, args ...string) error {
+	cmdObj := exec.Command(cmd, args...)
+
+	cmdObj.Dir = workDir
+	if err := os.MkdirAll(cmdObj.Dir, os.ModeDir|0755); err != nil {
```

Comment:
this is a helper function used only where the relevant step ends up invoking an external command - i don't think the mkdir belongs here because then 'who is responsible for mkdir' will be inconsistent between step types

Reply:
That makes sense, I moved the generation of the working directory into installrecipes. I'm not positive that's the best place for it so I'm open to suggestions as to where the directory should be made.